harness/determined/pytorch/_reducer.py

```diff
@@ -41,3 +42,146 @@ def _reduce_metrics(
+
+       -  Step 1 is fully defined by the user; it is not part of the interface.  This flexibility
+          allows for total flexibility when gathering individual values for reduction.
+
+       -  Step 2 is the MetricReducer.per_slot_reduce() interface.
+
+       -  Step 3 is the MetricReducer.cross_slot_reduce() interface.
+
+       -  The MetricReducer.reset() interface allows for MetricReducer reuse across many train and
+          validation workloads.
+
+    Example implementation and usage:
+
+    .. code:: python
+
+        class MyAvgMetricReducer(pytorch.MetricReducer):
+            def __init__(self):
+                self.reset()
+
+            def reset(self):
+                self.sum = 0
+                self.counts = 0
+
+            # User-defined mechanism for collecting values throughout
+            # training or validation. This update() mechanism demonstrates
+            # a computationally- and memory-efficient way to store the values.
+            def update(self, value):
+                self.sum += sum(value)
+                self.counts += 1
+
+            def per_slot_reduce(self):
+                # Because the chosen update() mechanism is so
+                # efficient, this is basically a noop.
+                return self.sum, self.counts
+
+            def cross_slot_reduce(self, per_slot_metrics):
+                # per_slot_metrics is a list of (sum, counts) tuples
+                # returned by the self.pre_slot_reduce() on each slot
+                sums, counts = zip(*per_slot_metrics)
+                return sum(sums) / sum(counts)
+
+
+        class MyPyTorchTrial(pytorch.PyTorchTrial):
+            def __init__(self, context):
+                # Register your custom reducer.
+                self.my_avg = context.wrap_reducer(
+                    MyAvgMetricReducer(), name="my_avg"
+                )
+                ...
+
+            def train_batch(self, batch, epoch_idx, batch_idx):
+                ...
+                # You decide how/when you call update().
+                self.my_avg.update(my_val)
+
+                # The "my_avg" metric will be included in the final
+                # metrics after the workload has completed; no need
+                # to return it here.
+                return {"loss": loss}
+
+
+    See also: :func:`determined.pytorch.PyTorchTrial.wrap_reducer`.
+    """
+
+    @abc.abstractmethod
+    def reset(self) -> None:
+        """
+        (This will be called automatically before any train or validation workload begins.)
+
+        Reset reducer state for another set of values.
+        """
+        pass
+
+    @abc.abstractmethod
+    def per_slot_reduce(self) -> Any:
+        """
+        (This will be called automatically after any train or validation workload ends.)
+
+        This will be called after all workers have finished (even when there is only one worker).
+
+        It should return some picklable value that is meaningful for cross_slot_reduce.
```

Comment:
No, it is definitely not an optional method.  Only the reduction calculation within this method is optional.  I removed the word "optional" from the docs entirely to avoid confusion.